pkg/detectors/npm/registry.go

```diff
@@ -0,0 +1,300 @@
+		}
+		return registryPat
+	}()
+)
+
+// findTokenRegistry returns the specific registry associated with the |token| if a high confidence match is found in |data|.
+//
+// Common configurations:
+// - npm: https://docs.npmjs.com/using-private-packages-in-a-ci-cd-workflow#create-and-check-in-a-project-specific-npmrc-file
+// - Yarn (TODO)
+// - Unity Package Manager (TODO)
+func findTokenRegistry(data string, token string) *registryInfo {
+	// .npmrc stores auth as `//registry.com/path/:authToken=$TOKEN
+	// Therefore, we should be able to correlate a token to a registry with a high degree of confidence.
+	registryAuthPat := regexp.MustCompile(fmt.Sprintf(
+		// language=regexp
+		`(?i)(//%s(?:/[a-z0-9._-]+)*)/:(?:_auth(?:Token)?|_password).{1,20}%s`, hostPat, token))
+	matches := registryAuthPat.FindStringSubmatch(data)
+	if len(matches) == 0 {
+		return nil
+	}
+
+	// A match was found, attempt to parse it.
+	uri := matches[1]
+	info := parseKnownRegistryURI(data, uri)
+	if info == nil {
+		info = parseUnknownRegistryURI(data, uri)
+	}
+	return info
+}
+
+// findAllRegistryURLs returns all instances of URLs that *look like* registries.
+// These are not associated with a specific token.
+func findAllRegistryURLs(data string) map[string]*registryInfo {
+	registries := make(map[string]*registryInfo)
+
+	// Look for known high-confidence matches.
+	for _, matches := range knownRegistryPat.FindAllStringSubmatch(data, -1) {
+		var (
+			_, uri = firstNonEmptyMatch(matches, 2) // first two matches are the entire string and protocol/prefix
+			info   = parseKnownRegistryURI(data, matches[1]+uri)
+		)
+		// Might be unnecessary, |info| is almost guaranteed to not be nil.
+		if info == nil {
+			continue
+		}
+		if _, ok := registries[info.Uri]; ok {
+			continue
+		}
+
+		registries[info.Uri] = info
+	}
+
+	// Attempt to parse any other low confidence matches.
+	for _, matches := range genericRegistryPat.FindAllStringSubmatch(data, -1) {
+		// Skip known registry patterns, those should have already been handled above.
+		if knownRegistryPat.MatchString(matches[0]) {
+			continue
+		}
+
+		_, uri := firstNonEmptyMatch(matches, 1) // first match is the entire string
+		info := &registryInfo{
+			RegistryType: other,
+		}
+		info.Scheme, info.Uri = parseRegistryURLScheme(data, uri)
+		if _, ok := registries[info.Uri]; ok {
+			continue
+		}
+
+		registries[info.Uri] = info
+	}
+
+	if len(registries) == 0 {
+		registries[defaultRegistryInfo.Uri] = defaultRegistryInfo
+	}
+	return registries
+}
+
+// parseKnownRegistryURI
+func parseKnownRegistryURI(data string, registryUri string) *registryInfo {
+	matches := knownRegistryPat.FindStringSubmatch(registryUri)
+	if len(matches) == 0 {
+		return nil
+	}
+
+	// Skip the first two indices: 1 is the entire string, 2 is the protocol.
+	index, uri := firstNonEmptyMatch(matches, 2)
+	info := &registryInfo{
+		RegistryType: registryType(index - 1),
```

Comment:
This is spooky black magic that relies on `knownRegistryPat` having the same number of capture groups as `registryType` has values. Unfortunately, it doesn't seem possible to programmatically get enum values so this can't be verified at compile time.

Reply:
Would it be feasible to create some sort of `registry` struct that contains a pattern and a name and a const int, and then smoosh them all together into a list in one place? That way there wouldn't have to be these multiple lists maintained across multiple files. (If not, then I think some loud comments in all the places that are magically correlated are warranted.)

Reply:
I considered doing that but was concerned that looping through _n_ patterns would be much less efficient. In the worst case, you'd need to loop through all the "known" patterns, find all "generic" registry matches, and do a secondary lookup to determine the scheme for each registry.

Reply:
Hm, I'm either misunderstanding something or being unclear. I'm just proposing merging the three lists of registries you have (const ints, string names, and patterns) into one list of structs, each of which contains three fields. Why would that add any additional loops?

Reply:
Perhaps I'm misunderstanding. Can you provide an example of what you mean?

Reply:
```
type registry struct {
    // maybe an int too, i don't know if it's necessary
    name    string
    pattern string
}

// somewhere, maybe at a top level
registries := []registry {
    {
        name: "npm",
        pattern: `(registry\.(?:npmjs\.org|yarnpkg\.com))`
    },
    {
        name: "artifactoryCloud",
        pattern: `([a-z0-9-]+\.jfrog\.io/(?:artifactory|[a-z0-9._-]+)/api/npm/[a-z][a-z0-9._-]+)`
    {,
    // etc
}
```

(This could also be some kind of map. I see that some of the patterns are pretty long and decomposable, so maybe they get built out of a function somehow.) Building the big monster pattern is a matter of looping through the slice and just joining all the patterns together. This would eliminate the spooky magic that ties together the big pattern and the registry list.